File: sample_source_code/JAVA/src/com/embold/sample/Shelter.java
package com.embold.sample;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

/**
 * Shelter holds a collection of animals.
 * Demonstrates composition: has-a relationship with IAnimal.
 * Multiple incoming calls from Report and Main.
 */
public class Shelter {

    private final String name;
    private final int capacity;
    private final List<IAnimal> animals;

    public Shelter(String name, int capacity) {
        this.name = name;
        this.capacity = capacity;
        this.animals = new ArrayList<>();
    }

    public String getName() { return name; }

    /** Adds an animal; returns {@code false} if at capacity. */
    public boolean intake(IAnimal animal) {
        if (animals.size() >= capacity) return false;
        animals.add(animal);
        return true;
    }

    /** Removes and returns the named animal, or empty if not found. */
    public Optional<IAnimal> adopt(String animalName) {
        Optional<IAnimal> found = animals.stream()
                .filter(a -> a.getName().equalsIgnoreCase(animalName))
                .findFirst();
        found.ifPresent(animals::remove);
        return found;
    }

    public List<IAnimal> getAnimals() {
        return Collections.unmodifiableList(animals);
    }

    public List<IAnimal> findBySpecies(String species) {
        return animals.stream()
                .filter(a -> a.getSpecies().equalsIgnoreCase(species))
                .collect(Collectors.toList());
    }

    public int availableCapacity() { return capacity - animals.size(); }

    public double averageAge() {
        return animals.stream().mapToInt(IAnimal::getAge).average().orElse(0.0);
    }
}

File: sample_source_code/python/src/shelter.py
"""Shelter: composition of IAnimal instances."""
from __future__ import annotations

from typing import Optional
from ianimal import IAnimal


class Shelter:
    """Manages a bounded collection of IAnimal instances (composition pattern)."""

    def __init__(self, name: str, capacity: int) -> None:
        self.name     = name
        self.capacity = capacity
        self._animals: list[IAnimal] = []

    def intake(self, animal: IAnimal) -> bool:
        """Add an animal; returns False if at capacity."""
        if len(self._animals) >= self.capacity:
            return False
        self._animals.append(animal)
        return True

    def adopt(self, animal_name: str) -> Optional[IAnimal]:
        """Remove and return the named animal, or None if not found."""
        for i, a in enumerate(self._animals):
            if a.name.lower() == animal_name.lower():
                return self._animals.pop(i)
        return None

    @property
    def animals(self) -> list[IAnimal]:
        return list(self._animals)

    def find_by_species(self, species: str) -> list[IAnimal]:
        return [a for a in self._animals if a.species.lower() == species.lower()]

    def available_capacity(self) -> int:
        return self.capacity - len(self._animals)

    def average_age(self) -> float:
        if not self._animals:
            return 0.0
        return sum(a.age for a in self._animals) / len(self._animals)
